feat(trends): Add moving average to trends

wmak opened the pull request:
- Trends will now use the same getInterval as discover but with high
  fidelity
- Use moving average instead of high intervals to smooth the graphs, and
  show a low opacity version of the noisy old graph

Before:
![image](https://user-images.githubusercontent.com/4205004/95638626-4a542280-0a63-11eb-81a6-a59997b94b26.png)

After:
<img width="853" alt="image" src="https://user-images.githubusercontent.com/4205004/95638905-5987a000-0a64-11eb-842d-e0e5b810e736.png">

- Window size of 10 was chosen experimentally.

2: Too noisy, almost matches non smoothed graph
![image](https://user-images.githubusercontent.com/4205004/95638747-bc2c6c00-0a63-11eb-9cc9-4d084210ad7f.png)

20: Smoothed, but doesn't look distinct enough from 10 for the data loss on either edge

![image](https://user-images.githubusercontent.com/4205004/95638794-e716c000-0a63-11eb-96e1-bda7b5ed8a65.png)

wmak committed:
feat(trends): Add moving average to trends

- Trends will now use the same getInterval as discover but with high
  fidelity
- Use moving average instead of high intervals to smooth the graphs, and
  show a low opacity version of the noisy old graph

benvinegar commented:
Looks nice. Showing both the smoothed and non-smoothed graphs gives me more confidence in the underlying data.

dashed commented on src/sentry/static/sentry/app/views/performance/trends/chart.tsx:
```diff
@@ -55,6 +61,38 @@ function transformEventStats(data: EventsStatsData, seriesName?: string): Series
+    } else if (i > currentData.length - AVERAGE_WINDOW / 2) {
+      value = lastValue;
+    } else {
+      value = movingAverage(currentData, i, AVERAGE_WINDOW);
```
I don't think this is how moving averages work. The curve for the moving average shouldn't be drawn until the 10th data point (`AVERAGE_WINDOW`) from the left.

wmak replied:
That is accurate, but I think it makes for a worse user experience since there's a gap at the beginning of every graph now:

![image](https://user-images.githubusercontent.com/4205004/95890458-d14e1700-0d51-11eb-9b50-4198c2d7a25b.png)

dashed replied:
Is there a possibility to do an extra fetch to paint these gaps? 😬

wmak replied:
🤔 there is, but i think that would add a lot of complexity, we'd have to update _only_ the event-stats call for trends to start an extra 20 x interval earlier (eg. 10 hours more for a 14d view). I'd rather have the gap, or use the first value for the gap.

markstory replied:
Could the gap at the beginning use partial data? It isn't ideal as the 'average' will line up with actual data too closely, but it might be better than a gap.

wmak replied:
:+1: Partial data looks much better than repeating the first value:
![image](https://user-images.githubusercontent.com/4205004/95909345-50504900-0d6c-11eb-83c4-d59bbf5a77a0.png)

dashed commented on src/sentry/static/sentry/app/views/performance/trends/utils.tsx:
```diff
@@ -408,6 +372,16 @@ export function getSelectedQueryKey(trendChangeType: TrendChangeType) {
+      .map(a => a.value)
+      .reduce((a, b) => a + b, 0) / size
+  );
+}
```
This is an odd way of calculating the moving average. 

For a given index `index`, the previous `size` data points should be used for calculating the average at the current index. The next `halfSize` data points shouldn't be used for the average.

markstory replied:
Agree, data from the 'future' shouldn't be included in a previous point's average. The average value for index=5 shouldn't include data from indexes 6-9.

wmak replied:
🤔 In hindsight I'm not sure why I did this, :+1: updating it to the expected calculation

markstory commented on src/sentry/static/sentry/app/views/performance/trends/chart.tsx:
```diff
@@ -55,6 +61,38 @@ function transformEventStats(data: EventsStatsData, seriesName?: string): Series
 
+function transformEventStatsSmoothed(data: Series[], seriesName?: string): Series[] {
+  const currentData = data[0].data;
+  const resultData = [] as SeriesDataUnit[];
```
```suggestion
  const resultData: SeriesDataUnit[] = [];
```
Should also work and doesn't use a cast operator.

markstory commented on src/sentry/static/sentry/app/views/performance/trends/chart.tsx:
```diff
@@ -55,6 +61,38 @@ function transformEventStats(data: EventsStatsData, seriesName?: string): Series
+    AVERAGE_WINDOW
+  );
+
+  for (let i = 0; i < currentData.length; i++) {
```
Should you start at the tail of the results so that the last window is complete? Right now you will have partial windows at the head and tail of the results. Alternatively you could start at the 10th element so that you're always working with full windows.

wmak replied:
So this ties back to Alberto's comments [here](https://github.com/getsentry/sentry/pull/21271#discussion_r503241557)
I wanted to avoid gaps at the beginning/end (now just beginning since moving average is correct) so i start at 0 to set the first average across the first 20 values.

wmak committed:
ref: Addressing PR comments

- Fixing movingAverage to use the last x points instead of last x/2 and
  next x/2
- Using better variable instantiation for resultData

wmak committed:
ref: Use a rougher moving average for the first points

markstory commented on src/sentry/static/sentry/app/views/performance/trends/chart.tsx:
```diff
@@ -64,12 +64,13 @@ function transformEventStats(data: EventsStatsData, seriesName?: string): Series
 
-  for (let i = 0; i < currentData.length; i++) {
-    let value = 0;
+  // Start at 1 so we don't ge NaN from the movingAverage
```
```suggestion
  // Start at 1 so we don't get NaN from the movingAverage
```

markstory commented on src/sentry/static/sentry/app/views/performance/trends/chart.tsx:
```diff
@@ -55,6 +61,32 @@ function transformEventStats(data: EventsStatsData, seriesName?: string): Series
+  const resultData: SeriesDataUnit[] = [];
+
+  // Start at 1 so we don't ge NaN from the movingAverage
+  for (let i = 1; i < currentData.length; i++) {
```
If you're starting at 1 will the moving average data be one result smaller than the unsmoothed data?

wmak replied:
Oh it is, I'll make the 0th point match the unsmoothed data's 0th point :+1:

wmak committed:
ref: Addressing PR comments

- fix spelling mistake
- add the 0th element to the resultData
